python-package/lightgbm/basic.py

```diff
@@ -682,6 +741,46 @@ def inner_predict(csr, num_iteration, predict_type, preds=None):
                 raise ValueError("Wrong length for predict results")
             return preds, nrow
 
+        def inner_predict_sparse(csr, num_iteration, predict_type):
+            ptr_indptr, type_ptr_indptr, __ = c_int_array(csr.indptr)
+            ptr_data, type_ptr_data, _ = c_float_array(csr.data)
+            csr_indices = csr.indices.astype(np.int32, copy=False)
+            matrix_type = C_API_MATRIX_TYPE_CSR
+            if type_ptr_indptr == C_API_DTYPE_INT32:
+                out_ptr_indptr = ctypes.POINTER(ctypes.c_int32)()
+            else:
+                out_ptr_indptr = ctypes.POINTER(ctypes.c_int64)()
+            out_ptr_indices = ctypes.POINTER(ctypes.c_int32)()
+            if type_ptr_data == C_API_DTYPE_FLOAT32:
+                out_ptr_data = ctypes.POINTER(ctypes.c_float)()
+            else:
+                out_ptr_data = ctypes.POINTER(ctypes.c_double)()
+            out_shape = np.zeros(2, dtype=np.int64)
+            _safe_call(_LIB.LGBM_BoosterPredictSparseOutput(
+                self.handle,
+                ptr_indptr,
+                ctypes.c_int32(type_ptr_indptr),
+                csr_indices.ctypes.data_as(ctypes.POINTER(ctypes.c_int32)),
+                ptr_data,
+                ctypes.c_int(type_ptr_data),
+                ctypes.c_int64(len(csr.indptr)),
+                ctypes.c_int64(len(csr.data)),
+                ctypes.c_int64(csr.shape[1]),
+                ctypes.c_int(predict_type),
+                ctypes.c_int(num_iteration),
+                c_str(self.pred_parameter),
+                ctypes.c_int(matrix_type),
+                out_shape.ctypes.data_as(ctypes.POINTER(ctypes.c_int64)),
+                ctypes.byref(out_ptr_indptr),
+                ctypes.byref(out_ptr_indices),
+                ctypes.byref(out_ptr_data)))
+            matrices = self.__create_sparse_native(csr, out_shape, out_ptr_indptr, out_ptr_indices, out_ptr_data,
+                                                   type_ptr_indptr, type_ptr_data, is_csr=True)
+            nrow = len(csr.indptr) - 1
+            return matrices, nrow
+
+        if predict_type == C_API_PREDICT_CONTRIB:
+            return inner_predict_sparse(csr, num_iteration, predict_type)
```

Comment:
Is possibility to predict feature contributions for matrices with `nrow > MAX_INT32` preserved?

Reply:
great question, why can't those native methods be changed to handle >MAX_INT32 rows for dense case?  For sparse case, only problem I see right now is the cast here for the passed in number of rows which can overflow:
int nrow = static_cast<int>(nindptr - 1);
and the code that follows this in PredictSparseCSR and CSC that stores row indexes.
However, I was just following the dense API style here.  I can change this but I don't quite understand why dense case has to do this batching and can't be modified to handle the full data.
Interestingly, dense case has this:
# __get_num_preds() cannot work with nrow > MAX_INT32, so calculate overall number of predictions piecemeal
but I don't call that method for sparse case.
I'm also not sure if this is really needed for computing local feature contributions, whereas I can understand this requirement more for predicting probabilities/values.

Reply:
I remember, this bunch of code with batching [was introduced](https://github.com/microsoft/LightGBM/pull/1615) exactly to implement [a feature request](https://github.com/microsoft/LightGBM/issues/1590) for supporting large data for feature contribution predictions.

And here is the reason behind `int32`:
> due to the efficiency concern, LightGBM only accept the int32 nrow.
For the nrow exceed the 32-bit ranges, please consider split it to multi data for prediction, and then concat back.
  https://github.com/microsoft/LightGBM/issues/1590#issuecomment-415066108

Reply:
that's crazy, that someone would want so many local feature importance values.  Why is using int32 instead of int64 nrow an efficiency concern?  How big is the perf impact?

Reply:
I think concatenating sparse matrices together into one from multiple is a bit more expensive than for dense data - eg you have to re-calculate the indptr

Reply:
maybe just for the sparse API I added we can accept >Int32 MAX values for nrow

Reply:
I'll update the PR with this fix

Reply:
> Why is using int32 instead of int64 nrow an efficiency concern? How big is the perf impact?

ping @guolinke

Reply:
@StrikerRUS I've fixed the code, actually I didn't need to change the external API as it was already accepting int64_t.  I think the fix is actually more trivial than I thought, and we should be able to accept CSR and CSC sparse matrices with more than int32 max rows.

Reply:
Great! Thank you very much!